pkg-orphan.1
.\" Copyright (C) 2005-2011 Victor Semionov <vsemionov@gmail.com>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions are met:
.\"  * Redistributions of source code must retain the above copyright notice,
.\"    this list of conditions and the following disclaimer.
.\"  * Redistributions in binary form must reproduce the above copyright notice,
.\"    this list of conditions and the following disclaimer in the documentation
.\"    and/or other materials provided with the distribution.
.\"  * Neither the name of the copyright holder nor the names of the contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
.\" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.\"
.TH "pkg-orphan" "1" "April 16, 2011" "FreeBSD" ""
.SH "NAME"
\fBpkg-orphan\fR -- manage unreferenced packages
.SH "SYNOPSIS"
\fBpkg-orphan\fR
[\fB-bBnpqr\fR]
[\fIname ...\fR]
.br 
\fBpkg-orphan\fR
{\fB-k\fR | \fB-l\fR | \fB-L\fR}
[\fB-pq\fR]
[\fIname ...\fR]
.br 
\fBpkg-orphan\fR
{\fB-a\fR | \fB-A\fR | \fB-d\fR}
[\fB-npq\fR]
\fIname ...\fR
.br 
\fBpkg-orphan\fR
{\fB-h\fR | \fB-v\fR}
.SH "DESCRIPTION"
There are three modes of operation - normal, listing, and keep-list update.

The default mode is normal. In this mode, orphan packages are selected for
removal, keeping or ignoring. The selection is done either by prompting
the user for input for each appropriate orphan package, or all of them
are selected for deletion or keeping (see batch mode below). The packages
that would be considered for deletion are those that: 1) are orphans;
2) match any given criteria; 3) are not present in the keep-list
(see keep-list below).

In normal mode, there are three possible user actions for detected orphans:
.IP
Delete - Delete the package (and its dependencies in recursive mode).
.br
Keep - Keep the package and add its name to the keep-list.
.br
Ignore - Ignore the package and don't add it to the keep-list.
.LP
Listing mode is used to show (all, or only a selection of) the list of orphans
or the list of kept packages (see below).

Keep-list update mode is used to add or remove package names from the
keep-list.

In each mode, the list of package names (or patterns), given on the command
line, limits the packages or entries that the mode would operate on. If a list
of names or patterns is not specified, every appropriate package or entry is
processed.

\fBKeep-list\fR

\fBpkg-orphan\fR maintains a file that contains a list of kept packages.
Its purpose is to avoid asking for the same package multiple times in
subsequent invocations. When detecting orphans in normal mode, if a
package's name is present in the keep-list, it automatically get skipped,
in both interactive and batch modes. When a considered package is selected
as "kept", its name is added to this list.

\fBMode switches\fR
.IP \fB-l\fR
List orphan packages.
.IP \fB-L\fR
Detailed list of orphan packages.
.IP \fB-k\fR
Print contents of the keep-list.
.IP \fB-a\fR
Add package names to the keep-list.
.IP \fB-A\fR
Same as \fB-a\fR, but limited to installed orphan package names.
.IP \fB-d\fR
Delete package names from the keep-list.
.IP \fB-h\fR
Print brief usage help and quit.
.IP \fB-v\fR
Print version and quit.
.SH "OPTIONS"
The following switches don't take arguments, but affect the behavior of
\fBpkg-orphan\fR:
.IP \fB-b\fR
Batch mode. Delete orphan packages, not present in the keep-list,
without asking.
.IP \fB-B\fR
Batch mode. Keep orphan packages and add them to the keep-list,
without asking.
.IP \fB-n\fR
Dry run. Don't delete or change anything, just show what would be done.
.IP \fB-q\fR
Be quiet. Only print errors and decision prompts when interactive.
.IP \fB-p\fR
Interpret \fIname ...\fR as a list of regular expressions.
This has no meaning with the \fB-a\fR switch.
.IP \fB-r\fR
Delete packages recursively. This is safe enough to not delete shared dependencies
(packages which are referenced by other orphans), or any previously kept packages,
that are currently not orphans.
.SH "TECHNICAL DETAILS"
To be as clear as possible, here is what happens internally in normal mode:

Package consideration:
.IP 1.
Prepare a list of all installed orphan packages.
.IP 2.
From this list, take only those packages which match the criteria,
supplied on the command line as a list of package names or patterns.
.IP 3.
Load the keep-list, if present.
.IP 4.
Mark for skipping all items in the list prepared in step 2,
that are present in the keep-list.
.IP 5.
For the rest of the items, either mark them for deletion or keeping
in batch mode, or ask the user for action in interactive mode.
.IP 6.
In recursive mode, for every orphan, marked for deletion, prepare a list
of all packages, referenced directly or indirectly by the
processed orphan.
.IP 7.
Do the actual deletion of the above orphan and its dependencies
(in recursive mode).

.LP
Package deletion (including dependencies):
.IP 1.
Abort removal if package name exists is keep-list. This is necessary
to prevent deletion of former orphans in the keep-list.
.IP 2.
Abort removal if the package is _currently_ an orphan. This prevents
deletion of shared dependencies.
.IP 3.
Call pkg_delete(1) to remove the package.
.SH "EXAMPLES"
.IP \fBo\fR
Check what would be done by recursively deleting all unkept orphans:
.br
pkg-orphan -brn
.IP \fBo\fR
Delete all unkept orphans, automatically:
.br
pkg-orphan -br
.IP \fBo\fR
Delete packages foo and bar:
.br
pkg-orphan -b foo bar
.IP \fBo\fR
Same as above:
.br
pkg-orphan -pb "^(foo|bar)$"
.IP \fBo\fR
Delete packages, whose name contains "baz":
.br
pkg-orphan -p "baz"
.IP \fBo\fR
Clear keep-list:
.br
pkg-orphan -pd ".*"
.IP \fBo\fR
See what would be done by the above:
.br
pkg-orphan -npd ".*"
.IP \fBo\fR
Add foo and bar to keep-list:
.br
pkg-orphan -a foo bar
.IP \fBo\fR
List all orphans with some details:
.br
pkg-orphan -L
.IP \fBo\fR
Show which kept packages start with "ba":
.br
pkg-orphan -pk "^ba"
.SH "ENVIRONMENT"
.IP PKG_DBDIR
Specifies an alternative location for the installed package database
and the kept package list.
.SH "FILES"
.IP /var/db/pkg/orphans.lst
This is the default location of the keep-list.
An alternate location can be given by setting the
PKG_DBDIR environment variable.
.SH "SEE ALSO"
pkg_info(1), pkg_delete(1)
.SH "AUTHORS"
Victor Semionov <vsemionov@gmail.com>


CHANGES
1.0.1:
- bug fix: when adding packages with -A, version numbers are no longer included in the keep list

0.1.1:
- small change to status messages, to resemble output from port builds
- changed email address

0.1.0:
- initial release


pkg-orphan.pl
#!/usr/bin/perl

# Copyright (C) 2005-2011 Victor Semionov <vsemionov@gmail.com>
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#  * Redistributions of source code must retain the above copyright notice,
#    this list of conditions and the following disclaimer.
#  * Redistributions in binary form must reproduce the above copyright notice,
#    this list of conditions and the following disclaimer in the documentation
#    and/or other materials provided with the distribution.
#  * Neither the name of the copyright holder nor the names of the contributors
#    may be used to endorse or promote products derived from this software
#    without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.


#use warnings;
use strict;

use Getopt::Std;


my $progname = "pkg-orphan";
my $version = "1.0.1";
my $copytight = "Copyright (C) 2005-2011 Victor Semionov <vsemionov\@gmail.com>";
my $rights = "All rights reserved.";

my $dbdir = (exists $ENV{PKG_DBDIR}? $ENV{PKG_DBDIR} : "/var/db/pkg");
my $keeplist = $dbdir . "/orphans.lst";
my %opts;

getopts("aAbBdhklLnpqrv", \%opts);

&one_switch("aAdhklLv");
&one_switch("bB");

if ($opts{v})
{
	&version();
}
elsif ($opts{h})
{
	&usage();
}
elsif ($opts{k})
{
	&load_kept(@ARGV);
	&list_kept();
}
elsif ($opts{l} || $opts{L})
{
	&load_orphans(@ARGV);
	&list_orphans();
}
elsif ($opts{a} || $opts{A})
{
	&load_kept();
	&add_kept($opts{A}, $opts{A}? &load_orphans(@ARGV? @ARGV : undef) : @ARGV);
	&save_kept();
}
elsif ($opts{d})
{
	&load_kept();
	&del_kept(@ARGV);
	&save_kept();
}
else
{
	&load_orphans(@ARGV);
	&load_kept();
	&sel_ops();
	&save_kept();
	&do_ops();
}





sub version
{
	my $help = "$progname $version\n$copytight\n$rights\n";
	print $help;
}

sub usage
{
	&version();

	my $help = "
Usage: $progname [-bBnpqr] [-a | -A | -d | -h | -k | -l | -L | -v] [name ...]

Without any switches, $progname enters interactive mode, asking what to do for
each unreferenced (orphan) package. Previously kept packages are skipped.

Commands:
  -a Add names to the keep-list.
  -A Same as -a, but limited to installed orphan package names.
  -d Delete names from the keep-list.
  -h Show this usage help.
  -k Show the keep-list.
  -l List orphan packages.
  -L Detailed list orphan packages.
  -v Show version.

Switches:
  -b Batch mode. Delete orphan packages, not present in the keep-list.
  -B Batch mode. Keep orphan packages and add them to the keep-list.
  -n Dry run. Don't delete or change anything, just show what would be done.
  -q Be quiet. Only print errors and decision prompts when interactive.
  -p Interpret <name> as a regular expression.
  -r Delete packages recursively. This will not delete shared dependencies.

";

	print $help;
}

sub get_packages
{
	my @list;
	my $pkg;
	chdir $dbdir or die "$0: Unable to change to directory $dbdir: $!\n";
	if (!opendir DBDIR, ".")
	{
		warn "$0: Unable to read directory $dbdir: $!\n";
		return ();
	}
	while ($pkg = readdir DBDIR)
	{
		if (($pkg ne ".") && ($pkg ne "..") && (-f "$pkg/+COMMENT"))
		{
			push @list, $pkg;
		}
	}
	close DBDIR;
	@list;
}

sub get_list
{
	my ($pkg, $file, $pat, $pre, $post) = @_;
	$file = "$dbdir/$pkg/$file";
	open FILE, $file or return undef;
	$pat = '.*' unless defined $pat;
	$pat = "$pre($pat)$post";
	my @list;
	while (<FILE>)
	{
		chomp;
		push @list, $1 if /$pat/;
	}
	close FILE;
	@list;
}

sub get_deps
{
	my $pkg = shift;
	&get_list($pkg, "+CONTENTS", '\S+', '^@pkgdep ');
}

sub get_reqs
{
	my $pkg = shift;
	&get_list($pkg, "+REQUIRED_BY", '\S+');
}

sub is_orphan
{
	my $pkg = shift;
	not int &get_reqs($pkg);
}

sub get_desc
{
	my $pkg = shift;
	(&get_list($pkg, "+COMMENT"))[0];
}

sub get_orphans
{
	my ($isre, @names) = @_;
	my @list;
	foreach my $pkg (&get_packages)
	{
		my $pn = &get_name($pkg);
		next if ( @names && !&grep_inverse($isre, $pn, \@names) );
		push @list, $pkg if &is_orphan($pkg);
	}
	@list;
}

sub ask_user
{
	my ($pkg, $num, $total) = @_;
	print "\n";
	print "Package $num of $total:\n";
	print "$pkg: ", &get_desc($pkg), "\n";
	print "$pkg: Delete/Keep/Ignore? [I]: ";
	my $ans = substr <STDIN>, 0, 1;
	print "\n";
	(int $ans =~ m/d/i, int $ans =~ m/k/i);
}

sub get_name
{
	my $pkg = shift;
	my ($name) = $pkg =~ /(.*)-.*/;
	$name;
}

sub in_names
{
	my ($pkg, $hash) = @_;
	exists $$hash{&get_name($pkg)};
}

sub grep_inverse
{
	my ($isre, $pat, $list) = @_;
	grep { $isre? $pat =~ /$_/ : $pat eq $_ } @$list;
}



my @orphans;
my %kept;
my %newkept;
my @ops;
my %deleted;

sub state
{
	print "===>  @_\n" unless $opts{q};
}

sub item
{
	my ($itm, $sgn) = @_;
	print "   [" . ($sgn? "+" : "-") . "] $itm\n" unless $opts{q};
}

sub load_orphans
{
	my @names = @_;
	&state("Loading list of orphaned packages.");
	@orphans = sort &get_orphans($opts{p}, @names);
}

sub list_orphans
{
	if ($opts{L})
	{
		&state("Listing orphaned packages (verbose).");
		print "$_: ", &get_desc($_), "\n" for (@orphans);
	}
	elsif ($opts{l})
	{
		&state("Listing orphaned packages.");
		print "$_\n" for (map &get_name($_), @orphans);
	}
	my $cnt = @orphans;
	&state("$cnt packages listed.");
}

sub load_kept
{
	my ($isre, @names) = ($opts{p}, @_);
	&state("Loading list of kept packages.");
	if (-f $keeplist)
	{
		open KEEP, $keeplist or warn "$0: Unable to read file $keeplist: $!\n";
		while (<KEEP>)
		{
			chomp;
			$kept{$_} = 1
				unless ( @names && !&grep_inverse($isre, $_, \@names) );
		}
		close KEEP;
	}
}

sub save_kept
{
	if (my @newnames = sort keys %newkept)
	{
		&add_kept(0, @newnames);
	}
	&state("Saving new list of kept packages.");
	unless ($opts{n})
	{
		if (open KEEP, ">$keeplist")
		{
			print KEEP "$_\n" foreach (sort keys %kept);
			close KEEP;
		}
		else
		{
			warn "$0: Unable to write file $keeplist: $!\n";
		}
	}
}

sub list_kept
{
	&state("Listing kept packages.");
	my @list = sort keys %kept;
	my $cnt = @list;
	print "$_\n" for (@list);
	&state("$cnt names listed.");
}

sub add_kept
{
	my $full = shift;
	my @names = @_;
	&state("Adding names to the list of kept packages.");
	my $cnt = 0;
	foreach my $pkg (@names)
	{
		my $name = (!$full? $pkg : &get_name($pkg));
		unless (exists $kept{$name})
		{
			&item($name, 1);
			$kept{$name} = 1;
			$cnt++;
		}
	}
	&state("$cnt names added.");
}

sub del_kept
{
	my ($isre, @names) = ($opts{p}, @_);
	&state("Removing names from the list of kept packages.");
	my $cnt = 0;
	foreach my $pkg (sort keys %kept)
	{
		if (&grep_inverse($isre, $pkg, \@names))
		{
			&item($pkg);
			delete $kept{$pkg};
			$cnt++;
		}
	}
	&state("$cnt names deleted.");
}

sub sel_ops
{
	&state("Selecting operations for packages.");
	my $unkept = @orphans - grep &in_names($_, \%kept), @orphans;
	for (my $i = my $idx = 0; $i<@orphans; $i++)
	{
		my $pkg = $orphans[$i];
		push @ops, 3;
		next if &in_names($pkg, \%kept);
		my ($del, $keep) = ($opts{b}, $opts{B});
		unless ($del || $keep)
		{
			($del, $keep) = &ask_user($pkg, ++$idx, $unkept);
		}
		$newkept{&get_name($pkg)} = 1 if $keep;
		$ops[-1] = ($del<<1) + $keep;
	}
}

sub do_ops
{
	&state("Performing selected operations on packages.");
	my @op_names = ("Ignoring", "Keeping ", "Deleting", "Skipping");
	my @cnt = (0, 0, 0, 0);
	foreach my $pkg (@orphans)
	{
		my $op = shift @ops;
		$cnt[$op]++;
		print "=> ", $op_names[$op], " $pkg\n" unless $opts{q};
		if ($op==2)
		{
			&pkg_delete($pkg, $opts{r}, $opts{n});
		}
	}
	my @dellist = values %deleted;
	my $ndel = grep { $_ != 0 } @dellist;
	my $nfail = @dellist - $ndel;
	my $norph = @orphans;
	&state("$norph considered: $ndel/$cnt[2] deleted, $cnt[1] kept, $cnt[3] skipped, $cnt[0] ignored; $nfail failed.");
}

sub wouldbe_orphan
{
	my $pkg = shift;
	not grep !exists $deleted{$_}, &get_reqs($pkg);
}

sub pkg_delete
{
	my ($pkg, $rec, $nop) = @_;
	my @list = $pkg;
	push @list, reverse &get_deps($pkg) if $rec;
	for (my $i=0; $i<@list; $i++)
	{
		my $cur = $list[$i];
		if (!$i || (!&in_names($cur, \%kept) && &wouldbe_orphan($cur)))
		{
			&item($cur);
			my $res = 0;
			$res = system "pkg_delete", $cur unless $nop;
			$deleted{$cur} = !$res;
		}
	}
}

sub one_switch
{
	my @arg = split(//, shift);
	my @con = grep { exists $opts{$_} } @arg;
	die "$0: Conflicting switches -$con[0] and -$con[1]\n"
		if (@con > 1);
}
